src/main/js/page_objects/blueocean/bluePipelineRunDetail.js

```diff
@@ -95,14 +117,55 @@ module.exports.commands = [{
         this.expect.element('@code').to.not.be.present.before(1000);
 
     },
+    validateGraph: function () {
+        var self = this;
+        self.waitForElementVisible('@highlightedGraph');
+        return self;
+    },
+    validateSteps: function (browser) {
+        var self = this;
+        self.waitForElementVisible('@logConsole');
+        browser.elements('css selector', '.logConsole', function (codeCollection) {
+            this.assert.equal(codeCollection.value.length > 0, true);
+        });
+        return self;
+    },
+    validateLog: function () {
+        var self = this;
+        self.waitForElementVisible('@code');
+        return self;
+    },
     validateEmpty: function () {
         var self = this;
         self.waitForElementVisible('@emptystate');
         return self;
     },
-    validateNotEmpty: function () {
+    validateNotEmptyArtifacts: function (browser, expectedCount) {
         var self = this;
         self.waitForElementVisible('@artifactTable');
+        if (browser && expectedCount) {
+            browser.elements('css selector', 'table.artifacts-table tr', function (codeCollection) {
+                this.assert.equal(codeCollection.value.length, expectedCount + 1); // +1 because of the heading row
+                return self;
+            });
+        }
         return self;
-    }
+    },
+    validateNotEmptyChanges: function (browser, expectedCount) {
+        var self = this;
+        self.waitForElementVisible('@changes');
+        if (browser && expectedCount) {
+            browser.elements('css selector', 'table.changeset-table tr', function (codeCollection) {
```

Comment:
elements is a selector that returns a collection of matching elements. This collection can then be analysed. 

The ```waitX``` only validate that the selector the use are returning true, However you cannot do any analysis about the elements this selectors are returiing.